packages/near-membrane-base/src/blue.ts

```diff
@@ -34,56 +35,37 @@ import {
     RedConstructor,
     RedFunction,
     RedValue,
+    ProxyTargetType,
+    BlueArrayOrObject,
+    RedArrayOrObject,
 } from './types';
 
-function renameFunction(provider: RedFunction, receiver: BlueFunction) {
-    try {
-        // a revoked proxy will break the membrane when reading the function name
-        const nameDescriptor = ReflectGetOwnPropertyDescriptor(provider, 'name')!;
-        ReflectDefineProperty(receiver, 'name', nameDescriptor);
-    } catch {
-        // intentionally swallowing the error because this method is just extracting the
-        // function in a way that it should always succeed except for the cases in which
-        // the provider is a proxy that is either revoked or has some logic to prevent
-        // reading the name property descriptor.
-    }
-}
-
 const ProxyCtor = Proxy;
 const { isArray: isArrayOrNotOrThrowForRevoked } = Array;
 
-function createBlueShadowTarget(target: BlueProxyTarget): BlueShadowTarget {
+function createShadowTarget(
+    targetTypeof: string,
+    targetIsArrowFunction: boolean,
+    targetFunctionName: string | undefined,
+    targetIsArray: boolean
+): BlueShadowTarget {
     let shadowTarget;
-    if (typeof target === 'function') {
+    if (targetTypeof === 'function') {
```

Comment:
@jdalton this is using string comparison, which we know it is going to be slow, but the good news is that we could refactor this to use numeric values since we only support two shapes, object or function.

Reply:
Yes. We could have `TYPE_OF_FUNCTION` which is say `1` and use that.

Reply:
> Yes. We could have `TYPE_ID_FUNCTION` which is say `1` and use that.

🤩  I love this idea.